File: patterns/Creational/AbstractFactory/Go/AbstractFactory.go
package main

import "fmt"

// Abstract Factory interface declares methods to create abstract products
type AbstractFactory interface {
	CreateProductA() AbstractProductA
	CreateProductB() AbstractProductB
}

// Concrete Factory 1 produces ProductA1 and ProductB1
type ConcreteFactory1 struct{}

func (f *ConcreteFactory1) CreateProductA() AbstractProductA {
	return &ConcreteProductA1{}
}

func (f *ConcreteFactory1) CreateProductB() AbstractProductB {
	return &ConcreteProductB1{}
}

// Concrete Factory 2 produces ProductA2 and ProductB2
type ConcreteFactory2 struct{}

func (f *ConcreteFactory2) CreateProductA() AbstractProductA {
	return &ConcreteProductA2{}
}

func (f *ConcreteFactory2) CreateProductB() AbstractProductB {
	return &ConcreteProductB2{}
}

// Abstract ProductA interface defines methods that ProductA must implement
type AbstractProductA interface {
	UsefulFunctionA() string
}

// Concrete ProductA1 implements the AbstractProductA interface
type ConcreteProductA1 struct{}

func (p *ConcreteProductA1) UsefulFunctionA() string {
	return "The result of the product A1."
}

// Concrete ProductA2 implements the AbstractProductA interface
type ConcreteProductA2 struct{}

func (p *ConcreteProductA2) UsefulFunctionA() string {
	return "The result of the product A2."
}

// Abstract ProductB interface defines methods that ProductB must implement
type AbstractProductB interface {
	UsefulFunctionB() string
	AnotherUsefulFunctionB(collaborator AbstractProductA) string
}

// Concrete ProductB1 implements the AbstractProductB interface
type ConcreteProductB1 struct{}

func (p *ConcreteProductB1) UsefulFunctionB() string {
	return "The result of the product B1."
}

func (p *ConcreteProductB1) AnotherUsefulFunctionB(collaborator AbstractProductA) string {
	result := collaborator.UsefulFunctionA()
	return fmt.Sprintf("The result of B1 collaborating with the (%s)", result)
}

// Concrete ProductB2 implements the AbstractProductB interface
type ConcreteProductB2 struct{}

func (p *ConcreteProductB2) UsefulFunctionB() string {
	return "The result of the product B2."
}

func (p *ConcreteProductB2) AnotherUsefulFunctionB(collaborator AbstractProductA) string {
	result := collaborator.UsefulFunctionA()
	return fmt.Sprintf("The result of B2 collaborating with the (%s)", result)
}

// Client code that works with the abstract factory and products
func clientCode(factory AbstractFactory) {
	productA := factory.CreateProductA()
	productB := factory.CreateProductB()

	fmt.Println(productB.UsefulFunctionB())
	fmt.Println(productB.AnotherUsefulFunctionB(productA))
}

func main() {
	fmt.Println("Client: Testing client code with the first factory type:")
	clientCode(&ConcreteFactory1{})

	fmt.Println("\nClient: Testing the same client code with the second factory type:")
	clientCode(&ConcreteFactory2{})
}

File: patterns/Creational/AbstractFactory/JavaScript/AbstractFactory.js
// AbstractFactory.js - Abstract Factory Interface
class AbstractFactory {
    createProductA() {
      throw new Error("Method 'createProductA()' must be implemented.");
    }
    
    createProductB() {
      throw new Error("Method 'createProductB()' must be implemented.");
    }
  }
  
  // ConcreteFactory1.js - Concrete Factory 1
  class ConcreteFactory1 extends AbstractFactory {
    createProductA() {
      return new ConcreteProductA1();
    }
    
    createProductB() {
      return new ConcreteProductB1();
    }
  }
  
  // ConcreteFactory2.js - Concrete Factory 2
  class ConcreteFactory2 extends AbstractFactory {
    createProductA() {
      return new ConcreteProductA2();
    }
    
    createProductB() {
      return new ConcreteProductB2();
    }
  }
  
  // AbstractProductA.js - Abstract Product A Interface
  class AbstractProductA {
    usefulFunctionA() {
      throw new Error("Method 'usefulFunctionA()' must be implemented.");
    }
  }
  
  // ConcreteProductA1.js - Concrete Product A1
  class ConcreteProductA1 extends AbstractProductA {
    usefulFunctionA() {
      return "The result of the product A1.";
    }
  }
  
  // ConcreteProductA2.js - Concrete Product A2
  class ConcreteProductA2 extends AbstractProductA {
    usefulFunctionA() {
      return "The result of the product A2.";
    }
  }
  
  // AbstractProductB.js - Abstract Product B Interface
  class AbstractProductB {
    usefulFunctionB() {
      throw new Error("Method 'usefulFunctionB()' must be implemented.");
    }
    
    anotherUsefulFunctionB(collaborator) {
      throw new Error("Method 'anotherUsefulFunctionB()' must be implemented.");
    }
  }
  
  // ConcreteProductB1.js - Concrete Product B1
  class ConcreteProductB1 extends AbstractProductB {
    usefulFunctionB() {
      return "The result of the product B1.";
    }
  
    anotherUsefulFunctionB(collaborator) {
      const result = collaborator.usefulFunctionA();
      return `The result of the B1 collaborating with the (${result})`;
    }
  }
  
  // ConcreteProductB2.js - Concrete Product B2
  class ConcreteProductB2 extends AbstractProductB {
    usefulFunctionB() {
      return "The result of the product B2.";
    }
  
    anotherUsefulFunctionB(collaborator) {
      const result = collaborator.usefulFunctionA();
      return `The result of the B2 collaborating with the (${result})`;
    }
  }
  
  // Client.js - Client Code
  function clientCode(factory) {
    const productA = factory.createProductA();
    const productB = factory.createProductB();
    
    console.log(productB.usefulFunctionB());
    console.log(productB.anotherUsefulFunctionB(productA));
  }
  
  // Usage
  console.log("Client: Testing client code with the first factory type:");
  clientCode(new ConcreteFactory1());
  
  console.log("\nClient: Testing the same client code with the second factory type:");
  clientCode(new ConcreteFactory2());
  